global_planner/src/planner_core.cpp

```diff
@@ -323,23 +323,40 @@ uint32_t GlobalPlanner::makePlan(const geometry_msgs::PoseStamped& start, const
       double best_sdist = DBL_MAX;
 
       unsigned int mx, my;
-      for(double dy = -tolerance; dy <= tolerance; dy += resolution){
-        p.pose.position.y = goal.pose.position.y + dy;
-        const double dx = std::sqrt(tolerance*tolerance - dy*dy);
-        for(p.pose.position.x = goal.pose.position.x - dx; p.pose.position.x <= goal.pose.position.x + dx; p.pose.position.x += resolution){
-            if(costmap_->worldToMap(p.pose.position.x, p.pose.position.y, mx, my)) {
-              unsigned int index = my * nx + mx;
-              double potential = potential_array_[index];
-              double sdist = sq_distance(p, goal);
-              if(potential < POT_HIGH && sdist < best_sdist){
-                best_sdist = sdist;
-                best_pose = p;
-                found_legal = true;
-                goal_blocked = false;
-              } else if (costmap_->getCost(goal_x_i, goal_y_i) < costmap_2d::INSCRIBED_INFLATED_OBSTACLE) {
-                goal_blocked = false;
+
+      // reduce tolerance to ensure all poses are clearly inside the tolerance
+      // TODO: use constexpr T max( std::initializer_list<T> ilist ); (since C++14)
+      const double tol_reduction = 1e-6 * std::max(std::max(1.0, tolerance), std::max(std::abs(p.pose.position.x), std::abs(p.pose.position.y)));
```

Comment:
Q: Can you explain better what's going on in this line? Why are we choosing the max of `{x, y, tolerance, 1.0}`? If you want to reduce the tolerance why not just `tolerance - 1e-6`? What's the purpose of the `max` here?

Reply:
This makes sure that the involved operations (subtraction from `tolerance`, addition to `goal.pose.position.x/y`) are significant regardless how large the involved numbers are.

For example:
```
    const double x = 1;
    std::cout << std::boolalpha << (x == x+1e-16) << std::endl;    
```
prints `true`. That is because `std::numeric_limits<double>::epsilon`, "the difference between 1.0 and the next value representable by the floating-point type", is `2.22045e-16`
https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

But yeah admittedly, just setting to 1e-6 would work for this particular use case, since position coordinates are limited to <1e3